### ISSUE TITLE
Memory initialization and updates

### ISSUE THREAD

**Issue by ozppupbg** [lucidrains/titans-pytorch#6]
I have some questions and possible issues regarding the memory handling.
1. From what I see, there is only one "MemoryMLP" per "NeuralMemory". Now, if training is done with multiple batches, shouldn't there be one "MemoryMLP" per batch? Otherwise the model would try to memorize all sequences in a batch into a single storage?
2. Maybe I'm missing something, but I could not find the "forgetting" mechanism in your code (the $\alpha_t$ / $\beta_t$ from equations 13 & 16)
3. If I understand correctly, the memory weights are initialized randomly in "memory_model". And then a zero initiated "state" is used for updates. From the equations mentioned above, I would expect the implementation in the paper to always start from zero memory weights, as the model could choose to erase the memory to zero. The idea of some "pretrained" memory could definitely be interesting, but it could be difficult to train in the presence of forgetting.
4. I think the state handling would not work, during inference, as https://github.com/lucidrains/titans-pytorch/blob/da5d4db662c7a52132027fec14f5e2f572a56d31/titans_pytorch/titans.py#L420 includes the `curr_weights` which would be added again in the next invocation https://github.com/lucidrains/titans-pytorch/blob/da5d4db662c7a52132027fec14f5e2f572a56d31/titans_pytorch/titans.py#L319
5. I could not yet figure out, how your memory weight updates correspond to equation 17. Where are the updates for each memory weight parameter calculated?

Thank you!

**Comment by lucidrains** [comment]
1. I'm doing per sample grads and batches are not reduced iirc
2. have this as an associative scan with the decay factor for now
3. yes that's a good point, I can make it an option, however, I think the network can also just learn to forget all of the past weights on the first few steps?
4. yeah, I haven't gotten to inference yet, to be done tomorrow 
5. it doesn't follow the equations, just hits on the major ideas

**Comment by rookie-joe** [comment]
i want to double-check that, u do not permanently modify the base weights of the memory model. right ?

instead u are doing like 
1. Calculates updates based on the input sequence
2. Adds these updates to the original weights temporarily during retrieval
3. Keeps track of these updates for future operations

which means, for each data example, the starting point is the same **during training**, am i right ?

**Comment by lucidrains** [comment]
@rookie-joe hi Jianqiao and thank you for bringing this up

yes, that was the case as of a few days ago. a memory model is learned and generates the surprise (gradients) for all timesteps, then accumulated by token dependent gates

however, i started to have doubts with either the paper (or perhaps my own understanding of it) that the model at each timestep should contain the accumulated gradients of all tokens before, at the time when the surprise is generated. that is what led to me adding a few bandaid solutions like adding previous neural memory layer (if not the first layer) surprises to the current one

i am about to go read the TTT paper in more detail to figure out what i'm missing here, because that doesn't seem quite right either. any help would be greatly appreciated

**Comment by lucidrains** [comment]
ah yea, all the missing details in this paper is found in the TTT paper. titans is an incremental paper built on top of TTT

the momentum is the only real contribution

**Comment by lucidrains** [comment]
the day when an LLM can reliably do latex sequence to python sequence cannot come soon enough

**Comment by lucidrains** [comment]
ah yea, TTT explains all that was missing in the Titans paper. what i have now amounts to a single minibatch, but not across batches

**Comment by jeremygatineau** [comment]
I also just took a another look at [TTT](https://arxiv.org/pdf/2407.04620) and yeah it does seem like it's per-layer, i.e. the $x_t$ stands in as the layer's input not the whole model's input. At the same time I don't think Titans or TTT people reported trying the gradient accumulation approach so there may be a chance that it's actually better. 

In that case I guess the next question is whether the backprop of memory module $M_t^i$ at layer $i$ w.r.t. the true model input $x_t$ (so the 0th embedded sequence $h^0_t$) yielding $M_{t+1}^i = M_{t}^i - \nabla l(M_t^i ; h^0_t)$ (instead of $M_{t+1}^i = M_{t}^i - \nabla l(M_t^i ; h^i_t)$) reduces to the gradient accumulation (or maybe a multiplicative version of it), which, at a glance would only be the case if the memory modules only depended on one isolated memory backbone, otherwise we would also need the gradients of attention branch, or to fully compute the gradients w.r.t $h^0_t$ which seems very expensive.

**Comment by jeremygatineau** [comment]
Also, and you probably saw this already, but this seems quite relevant 

![Image](https://github.com/user-attachments/assets/a1342a59-d240-45a4-a246-4481dd3ade96)

**Comment by lucidrains** [comment]
@jeremygatineau yea, i'm currently reading the TTT paper in detail for the first time. most of the unanswered questions in Titans are actually in the TTT paper

would you like to join a Signal group chat to discuss this? could use any help with this rather new line of research

**Comment by jeremygatineau** [comment]
Sure!

**Comment by rookie-joe** [comment]
Dear Sir,

I would like to discuss concerns regarding gradient updates in our current setup.

Typically, we use one gradient tensor with the same shape as the parameters. For example, if the memory network is a single-layer MLP, we assume the shape is `2 * [d, d]` (up and down). This `[d, d]` gradient is all-reduced across a global batch to update the weights. 

However, in this case, suppose the input shape is `[b, L, d]`, and the chunk size is `c`. We would then need gradients for each batch and each chunk, resulting in GPU memory usage of `2 * [b, L//c, d, d]` for storing gradients. Additionally, the memory required to calculate the gradients can reach `2 * c * d * d`. 

For example:
- Batch size (`b`) = 30
- Sequence length (`L`) = 8192
- Chunk size (`c`) = 512
- Hidden dimension (`d`) = 1024

The memory usage for each layer would be approximately:

```
2 * c * d * d + 2 * [b, L//c, d, d] ≈ 1.5GB (bfloat16
```
which is far more bigger than model weights + input hidden states. and this is only the inner loop gradient usage. Moreover, this doesn't account for 1. the more deeper mlp 2. the need to store gradients across layers and 3.the other additional hidden states like momentum, function call during retrieval (which not update weights but still need to get cur_weight), etc, which further increases the memory footprint. 

To address this, I think we might need a mechanism to update gradients into weights incrementally during chunk-wise retrieval/updates, which could help reduce GPU memory stress. 

I would greatly appreciate your feedback on this matter.

**Comment by lucidrains** [comment]
@rookie-joe  yea i know, that is why they needed to resort to chunking

perhaps an MLP with [factorized weights](https://github.com/lucidrains/titans-pytorch/blob/main/titans_pytorch/memory_models.py#L117) could be tried?

**Comment by rookie-joe** [comment]
> [@rookie-joe](https://github.com/rookie-joe) yea i know, that is why they needed to resort to chunking
> 
> perhaps an MLP with [factorized weights](https://github.com/lucidrains/titans-pytorch/blob/main/titans_pytorch/memory_models.py#L117) could be tried?

thks, let me take a look at  [factorized weights](https://github.com/lucidrains/titans-pytorch/blob/main/titans_pytorch/memory_models.py#L117), btw i am also interested in TTT and the further linear transformers, can u help add me to the Signal group chat to discuss this ?

**Comment by lucidrains** [comment]
@rookie-joe yea! hit me up on Signal
